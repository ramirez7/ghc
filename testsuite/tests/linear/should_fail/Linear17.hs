{-# LANGUAGE LinearTypes #-}
{-# LANGUAGE UnicodeSyntax #-}
{-# LANGUAGE RebindableSyntax #-}
module Linear17 where

-- Rebindable do notation

(>>=) :: a ⊸ (a ⊸ b) ⊸ b
(>>=) x f = f x

-- Not sure why `fail` and return are needed in my examples
fail :: a
fail = error "fail"

return :: a ⊸ a
return x = x

incorrectDo1 = do
  x <- ()
  (y,z) <- ((),())
  () <- y
  () <- z
  ()

incorrectDo2 = do
  x <- ()
  (y,z) <- ((),x)
  () <- y
  ()

incorrectDo3 = do
  x <- ()
  (y,z) <- (x,x)
  () <- y
  () <- z
  ()
