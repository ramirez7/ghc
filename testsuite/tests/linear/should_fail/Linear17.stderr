
Linear17.hs:17:3: error:
    • Couldn't match expected multiplicity ‘'One’ of variable ‘x’ with actual multiplicity ‘'Omega’
    • In a stmt of a 'do' block: x <- ()
      In the expression:
        do x <- ()
           (y, z) <- ((), ())
           () <- y
           () <- z
           ....
      In an equation for ‘incorrectDo1’:
          incorrectDo1
            = do x <- ()
                 (y, z) <- ((), ())
                 () <- y
                 ....

Linear17.hs:25:6: error:
    • Couldn't match expected multiplicity ‘'One’ of variable ‘z’ with actual multiplicity ‘'Omega’
    • In the pattern: (y, z)
      In a stmt of a 'do' block: (y, z) <- ((), x)
      In the expression:
        do x <- ()
           (y, z) <- ((), x)
           () <- y
           ()

Linear17.hs:30:3: error:
    • Couldn't match expected multiplicity ‘'One’ of variable ‘x’ with actual multiplicity ‘'Omega’
    • In a stmt of a 'do' block: x <- ()
      In the expression:
        do x <- ()
           (y, z) <- (x, x)
           () <- y
           () <- z
           ....
      In an equation for ‘incorrectDo3’:
          incorrectDo3
            = do x <- ()
                 (y, z) <- (x, x)
                 () <- y
                 ....
